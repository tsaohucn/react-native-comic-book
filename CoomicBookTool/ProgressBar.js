import React, { Component } from 'react'
import { 
  View, 
  StyleSheet, 
  Text, 
  Animated, 
  Dimensions, 
  TouchableWithoutFeedback 
} from 'react-native'
import { Icon } from 'react-native-elements'
import { Slider } from 'react-native-elements'


export default class ProgressBar extends Component {

  constructor(props) {
    super(props)
    this.pageNumber = null
    this.previousChapterPageNumber = null
    this.nextChapterPageNumber = null
    this.state = {
      value: 0
    }
  }

  receivePageNumber = pageNumber => {
    this.getChapterPageNumber(pageNumber)
  }

  getChapterPageNumber = pageNumber => {
    const chapterPageNumbers = this.props.chapter.map(ele => ele.pageNumber)
    const previousChapterPageNumbers = chapterPageNumbers.filter(_pageNumber => _pageNumber < pageNumber)
    const nextChapterPageNumbers = chapterPageNumbers.filter(_pageNumber => _pageNumber > pageNumber)
    this.previousChapterPageNumber = previousChapterPageNumbers[previousChapterPageNumbers.length - 1]
    this.nextChapterPageNumber = nextChapterPageNumbers[0]
    this.setState({
      value: pageNumber - 1
    })
  }

  onClickPreviousChapter = () => {
    this.props.onClickPreviousChapter && this.props.onClickPreviousChapter(this.previousChapterPageNumber)
  }

  onClickNextChapter = () => {
    this.props.onClickNextChapter && this.props.onClickNextChapter(this.nextChapterPageNumber)
  }

  onValueChange = value => {
    this.setState({value})
  }

  onSlidingComplete = value => {
    const pageNumber = value + 1
    this.props.onProgressComplete && this.props.onProgressComplete(pageNumber)
  }

  render() {
    return(
      <Animated.View style={[styles.animated,{
        transform: [
          { translateY: this.props.animatedProgressBarY }
        ]}]}
      >
          <View style={styles.columnOne}>
            <Text style={styles.text}>{this.state.value + 1 + '/' + this.props.finalPageNumber}</Text>
          </View>
          <View style={styles.columnTwo}>
            <TouchableWithoutFeedback
              onPress={this.onClickPreviousChapter}
            >
              <View style={styles.iconView}>
                <Icon
                  name='page-first'
                  type='material-community'
                  color='white'
                />
              </View>
            </TouchableWithoutFeedback>
              <Slider 
                onSlidingStart={this.onSlidingStart}
                onValueChange={this.onValueChange} 
                onSlidingComplete={this.onSlidingComplete}
                value={this.state.value}
                step={1}
                maximumValue={this.props.finalPageNumber - 1}
                style={styles.slider}
                minimumTrackTintColor={'white'}
                maximumTrackTintColor={'grey'}
                thumbTintColor={'white'}
              />
            <TouchableWithoutFeedback
              onPress={this.onClickNextChapter}
            >
              <View style={styles.iconView}>
                <Icon
                  name='page-last'
                  type='material-community'
                  color='white'
                />
              </View>
            </TouchableWithoutFeedback>
          </View>
      </Animated.View>
    )
  }

}

const { width, height } = Dimensions.get('window')

const styles = StyleSheet.create({
  animated: {
    position: 'absolute',
    bottom: 0,
    width,
    height: 100,   
    backgroundColor: 'rgba(52, 52, 52, 0.8)'
  },
  columnOne: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center'
  },
  columnTwo: {
    flex: 1,
    flexDirection: 'row',
    alignItems: 'center'
  },
  slider: {
    width: width*5/7
  },
  iconView: {
    flex: 1
  },
  text: {
    fontSize: 20,
    color: 'white',
    padding: 10
  }
})