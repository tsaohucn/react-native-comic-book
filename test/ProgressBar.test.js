// __tests__/ProgressBar.js
import 'react-native'
import React from 'react'
import ProgressBar from '../CoomicBookTool/ProgressBar'

import renderer from 'react-test-renderer'

jest.mock('react-native-elements', () => {
	return {
		Icon: 'Icon',
		Slider: 'Slider'
	}
})

describe('Testing ProgressBar component', () => {

	it('renders correctly with no props', () => {
		const props = {

		}
	  const tree = renderer.create(
	    <ProgressBar {...props}/>
	    ).toJSON();
	  expect(tree).toMatchSnapshot();
	})

	it('renders correctly with animatedProgressBarY value', () => {
		const props = {
			animatedProgressBarY: 100
		}
	  const tree = renderer.create(
	    <ProgressBar {...props}/>
	    ).toJSON();
	  expect(tree).toMatchSnapshot();
	})

/*
	it('test component method', () => {

	  const wrapper = renderer.create(
	    <ProgressBar/>
	   )
		wrapper.instance().onValueChange()

	})*/

})