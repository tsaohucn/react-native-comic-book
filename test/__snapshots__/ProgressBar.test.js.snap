// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing ProgressBar component renders correctly with animatedProgressBarY value 1`] = `
<View
  collapsable={undefined}
  style={
    Object {
      "bottom": 0,
      "height": 100,
      "position": "absolute",
      "transform": Array [
        Object {
          "translateY": 100,
        },
      ],
      "width": 750,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "rgba(52, 52, 52, 0.8)",
          "borderRadius": 15,
          "height": 40,
          "justifyContent": "center",
          "left": 107.14285714285714,
          "position": "absolute",
          "width": 80,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "white",
            "fontSize": 15,
          }
        }
      >
        1/undefined
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "rgba(52, 52, 52, 0.8)",
        "flex": 1,
        "flexDirection": "row",
      }
    }
  >
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      hitSlop={undefined}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "flex": 1,
        }
      }
      testID={undefined}
    >
      <Icon
        color="white"
        name="page-first"
        size={30}
        type="material-community"
      />
    </View>
    <Slider
      maximumTrackTintColor="grey"
      maximumValue={NaN}
      minimumTrackTintColor="white"
      onSlidingComplete={[Function]}
      onSlidingStart={undefined}
      onValueChange={[Function]}
      step={1}
      style={
        Object {
          "justifyContent": "center",
          "width": 535.7142857142857,
        }
      }
      thumbStyle={
        Object {
          "transform": Array [
            Object {
              "translateY": -1.5,
            },
          ],
        }
      }
      thumbTintColor="white"
      trackStyle={
        Object {
          "height": 1,
          "justifyContent": "center",
        }
      }
      value={0}
    />
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      hitSlop={undefined}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "flex": 1,
        }
      }
      testID={undefined}
    >
      <Icon
        color="white"
        name="page-last"
        size={30}
        type="material-community"
      />
    </View>
  </View>
</View>
`;

exports[`Testing ProgressBar component renders correctly with no props 1`] = `
<View
  collapsable={undefined}
  style={
    Object {
      "bottom": 0,
      "height": 100,
      "position": "absolute",
      "transform": Array [
        Object {
          "translateY": undefined,
        },
      ],
      "width": 750,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "rgba(52, 52, 52, 0.8)",
          "borderRadius": 15,
          "height": 40,
          "justifyContent": "center",
          "left": 107.14285714285714,
          "position": "absolute",
          "width": 80,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "white",
            "fontSize": 15,
          }
        }
      >
        1/undefined
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "rgba(52, 52, 52, 0.8)",
        "flex": 1,
        "flexDirection": "row",
      }
    }
  >
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      hitSlop={undefined}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "flex": 1,
        }
      }
      testID={undefined}
    >
      <Icon
        color="white"
        name="page-first"
        size={30}
        type="material-community"
      />
    </View>
    <Slider
      maximumTrackTintColor="grey"
      maximumValue={NaN}
      minimumTrackTintColor="white"
      onSlidingComplete={[Function]}
      onSlidingStart={undefined}
      onValueChange={[Function]}
      step={1}
      style={
        Object {
          "justifyContent": "center",
          "width": 535.7142857142857,
        }
      }
      thumbStyle={
        Object {
          "transform": Array [
            Object {
              "translateY": -1.5,
            },
          ],
        }
      }
      thumbTintColor="white"
      trackStyle={
        Object {
          "height": 1,
          "justifyContent": "center",
        }
      }
      value={0}
    />
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      hitSlop={undefined}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "flex": 1,
        }
      }
      testID={undefined}
    >
      <Icon
        color="white"
        name="page-last"
        size={30}
        type="material-community"
      />
    </View>
  </View>
</View>
`;
